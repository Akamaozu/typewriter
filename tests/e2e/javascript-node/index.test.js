/* eslint-disable no-null/no-null */
/* eslint-disable @typescript-eslint/camelcase */
import {
	emptyEvent,
	eventWithAllTypes,
	I42TerribleEventName3,
	exampleNamingCollision,
	exampleNamingCollision1,
	violationHandlerTest,
	setTypewriterOptions,
} from './analytics'
import SegmentAnalytics from 'analytics-node'
import fetch from 'node-fetch'
import { promisify } from 'util'

const SIDECAR_ADDRESS = 'http://localhost:8765'

test('Validate Analytics Calls', async cb => {
	// Initialize an analytics-node instance.
	const analytics = new SegmentAnalytics('123456', {
		host: SIDECAR_ADDRESS,
	})
	analytics.flush = promisify(analytics.flush)

	const userId = 'user-1234'

	// Verify analytics calls fail before setting an analytics instance.
	expect(() => {
		emptyEvent({
			userId,
		})
	}).toThrow()

	// Initialize Typewriter with an analytics-node instance.
	setTypewriterOptions({
		analytics,
	})

	// Send an event with no properties.
	emptyEvent({
		userId,
	})

	// Send an event with every combination of property type.
	eventWithAllTypes({
		properties: {
			'required any': 123,
			'required array': [
				{
					'required sub-property': 'Hello World',
				},
			],
			'required array (empty)': [123, 'Hello World'],
			'required boolean': false,
			'required int': 123,
			'required nullable string': null,
			'required number': 3.1415,
			'required number or string': 123,
			'required object': {
				'required sub-property': 'Hello World',
			},
			'required object (empty)': {},
			'required string': 'Hello World',
			'required string regex': 'FOO',
		},
		userId,
	})

	// Test an event with poor naming standards, to validate sanitization.
	I42TerribleEventName3({
		properties: {
			'0000---terrible-property-name~!3': 'foobar',
			propertyNameCollision: 'camelcase',
			property_name_collision: 'snakecase',
		},
		userId,
	})

	exampleNamingCollision({
		userId,
	})
	exampleNamingCollision1({
		userId,
	})

	// TODO: Test the handler for unknown methods.

	// Test the onViolation error handler.
	// By default, it'll throw an error on Violation in NODE_ENV=test.
	expect(() => {
		violationHandlerTest({
			properties: {
				'required string': 123,
			},
			userId,
		})
	}).toThrow('You made an analytics call (Violation Handler Test)')

	// Verify that a custom onViolation error handler is called on Violation.
	setTypewriterOptions({
		onViolation: (msg, violations) => {
			expect(msg.event).toEqual('Violation Handler Test')
			expect(violations).toHaveLength(1)

			throw new Error('onViolation called')
		},
	})
	expect(() => {
		violationHandlerTest({
			properties: {
				'required string': 123,
			},
			userId,
		})
	}).toThrow('onViolation called')

	// Verify that onViolation error handler is not called unless there is a Violation.
	violationHandlerTest({
		properties: {
			'required string': 'Hello World',
		},
		userId,
	})

	await analytics.flush()

	const resp = await fetch(`${SIDECAR_ADDRESS}/messages`)
	const messages = await resp.json()

	expect(messages).toMatchSnapshot()

	cb()
})
