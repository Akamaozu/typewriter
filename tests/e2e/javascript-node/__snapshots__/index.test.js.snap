// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Analytics Calls 1`] = `
Array [
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "Empty Event",
    "properties": Object {},
    "type": "track",
    "userId": "user-1234",
  },
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "Event with All Types",
    "properties": Object {
      "required any": 123,
      "required array": Array [
        Object {
          "required sub-property": "Hello World",
        },
      ],
      "required array (empty)": Array [
        123,
        "Hello World",
      ],
      "required boolean": false,
      "required int": 123,
      "required nullable string": null,
      "required number": 3.1415,
      "required number or string": 123,
      "required object": Object {
        "required sub-property": "Hello World",
      },
      "required object (empty)": Object {},
      "required string": "Hello World",
      "required string regex": "FOO",
    },
    "type": "track",
    "userId": "user-1234",
  },
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "42_--terrible==\\"event'++name~!3",
    "properties": Object {
      "0000---terrible-property-name~!3": "foobar",
      "propertyNameCollision": "camelcase",
      "property_name_collision": "snakecase",
    },
    "type": "track",
    "userId": "user-1234",
  },
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "example naming collision",
    "properties": Object {},
    "type": "track",
    "userId": "user-1234",
  },
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "example_naming_collision",
    "properties": Object {},
    "type": "track",
    "userId": "user-1234",
  },
  Object {
    "_metadata": Object {
      "nodeVersion": "10.15.3",
    },
    "context": Object {
      "library": Object {
        "name": "analytics-node",
        "version": "3.3.0",
      },
      "typewriter": Object {
        "language": "javascript",
        "version": "7.0.0",
      },
    },
    "event": "Violation Handler Test",
    "properties": Object {
      "required string": "Hello World",
    },
    "type": "track",
    "userId": "user-1234",
  },
]
`;
